csrc/id_graphs.cpp

```diff
@@ -0,0 +1,3453 @@
+  //
+  // The almost exact map will map T1's and T2's b1 together, but they're being
+  // resolved to i1 and i2 respectively. So we want to have separate entries so
+  // we can have an easy to process promotion map.
+  //
+  // Loop is a permissive like map, it could have many entries, use the exact
+  // map as the one we iterate on to reduce complexity as it hopefully has
+  // smaller groups and this algorithm scales with the number of groups *
+  // (number of entries in groups ^ 2)
+
+  auto intersection_exact_loop_graph = initializeIdGraph();
+  for (auto exact_group :
+       idGraph(IdMappingMode::EXACT).disjointIdSets().disjointSets()) {
+    auto set_size = exact_group->size();
+    for (auto id0_i : c10::irange(set_size)) {
+      auto id0 = exact_group->vector()[id0_i];
+      for (auto id1_i = id0_i; id1_i < set_size; id1_i++) {
+        auto id1 = exact_group->vector()[id1_i];
+        // id0 and id1 map in the almost exact map, if they also map in the loop
+        // graph, then add the mapping to the inersection
+        if (idGraph(IdMappingMode::LOOP)
+                .disjointIdSets()
+                .strictAreMapped(id0, id1)) {
+          intersection_exact_loop_graph.mapIds(id0, id1);
+        }
+      }
+    }
+  }
+
+  // std::cout << "Intersection exact - loop: " << std::endl;
+  // {
+  //   IdGroups groups;
+  //   for (auto group :
+  //        intersection_exact_loop_graph.disjointIdSets().disjointSets()) {
+  //     groups.pushBack(group);
+  //   }
+  //   std::cout << debug_print::idGroupsStringShort(groups) << std::endl;
+  // }
+
+  // Promotion logic is going to be on the intersection of the exact and loop
+  // graph. We will generate a map on the entries of this graph so it's
+  // important to not modify this graph moving forward, as that would invalidate
+  // the map.
+  //
+  // iel stands for Intersection of the Exact and Loop graphs.
+  std::unordered_map<IdGroup, IterDomain*> iel_promotion_map;
+
+  // This should probably work just on terminating inputs, as we shouldn't be
+  // able to modify a broadcast domain between root and rfactor which would be
+  // required to resolve a non input broadcast domain. But for now leaving it as
+  // traversal on all broadcast groups.
+  for (auto iel_group :
+       intersection_exact_loop_graph.disjointIdSets().disjointSets()) {
+    if (!iel_group->front()->isBroadcast()) {
+      continue;
+    }
+
+    // Collect all the exact groups of the resolutions of the broadcast id's
+    IdGroups resolved_exact_groups;
+    for (auto bcast_id : *iel_group) {
+      auto p2c_root_broadcast_resolution_map_it =
+          p2c_root_broadcast_resolution_map.find(bcast_id);
+
+      if (p2c_root_broadcast_resolution_map_it ==
+          p2c_root_broadcast_resolution_map.end()) {
+        continue;
+      }
+
+      resolved_exact_groups.pushBack(
+          idGraph(IdMappingMode::EXACT)
+              .toGroups(p2c_root_broadcast_resolution_map_it->second));
+    }
+
```

Comment:
I actually think we could support a case where it resolves to multiple IdGroups, but that would require changes in the loop extent and predicates. As the loop extent would then have to be the max of the two broadcasts.

```
// Broadcast and concretize same domain in two different ways and try to merge
// their loops remains unsupported.
TEST_F(NVFuserTest, FusionMultiPromotion2_CUDA) {
```

Should be safe to add the assert and reference this test in case we want to change in the future.

Reply:
Oh, read further in the code, yeah we can pull that assert up.

Reply:
Sorry, my understanding when writing the above was not quite right...

> So what we're just doing is trying to map broadcast IEL IdGroups to their concretized IEL groups, if that concretization is performed within the loop group. If it's not performed within the loop group then it doesn't need to be replayed as if it was.

There could be multiple entries in `p2c_root_broadcast_resolution_map`, that's fine. We can only support one of those within the loop group. i.e. we can resolve a broadcast to many extents within a kernel, but for now only one of those resolutions can show up within a loop group. We're looking for the instance that's within the loop group.

`p2c_root_broadcast_resolution_map` is all resolutions in the fusion/kernel.

Reply:
Do you mean `FusionMultiPromotion2` has two exact resolved groups? The test seems have a couple of bugs, which I commented about. After fixing them, I don't see multiple groups. Here's the output I'm seeing with some more printf I added:

```
IEL: { bS7{1} }
Resolved exact groups:
{ iS24{T1.size[1]}; iS22{T1.size[1]} }
Num resolved groups: 1
IEL: { bS11{1} }
Resolved exact groups:
{ iS29{T2.size[1]}; iS27{T2.size[1]} }
Num resolved groups: 1
Initial promotion map:
  iS24{T1.size[1]} <- { bS7{1} }
  iS29{T2.size[1]} <- { bS11{1} }
```

Reply:
I believe this is now fixed.